AmplifyPlugins/API/Sources/AWSAPIPlugin/AppSyncRealTimeClient/AppSyncRealTimeClient.swift

```diff
@@ -0,0 +1,457 @@
+            log.debug("[AppSyncRealTimeClient] start subscription failed, could not found subscription with id \(id) ")
+            throw APIError.unknown("Could not find a subscription with id \(id)", "", nil)
+        }
+
+        try await subscription.subscribe(
+            with: webSocketClient,
+            requestInterceptor: requestInterceptor,
+            responseStream: subject.eraseToAnyPublisher()
+        )
+
+        return AnyCancellable {
+            Task { [weak self] in
+                guard let self else { return }
+                try await subscription.unsubscribe(
+                    with: self.webSocketClient,
+                    responseStream: self.subject.eraseToAnyPublisher()
+                )
+            }
+        }
+
+    }
+
+    private func subscribeToWebSocketEvent() async {
+        await self.webSocketClient.publisher.sink { [weak self] _ in
+            self?.log.debug("[AppSyncRealTimeClient] WebSocketClient terminated")
+        } receiveValue: { webSocketEvent in
+            Task { [weak self] in
+                await self?.onWebSocketEvent(webSocketEvent)
+            }
+        }
+        .store(in: &cancellables)
+    }
+
+    private func resumeExistingSubscriptions() {
+        log.debug("[AppSyncRealTimeClient] Resuming existing subscriptions")
+        for (id, _) in self.subscriptions {
+            Task {
+                do {
+                    try await self.startSubscription(id).store(in: &cancellablesBindToConnection)
+                } catch {
+                    log.debug("[AppSyncRealTimeClient] Failed to resume existing subscription with id: (\(id))")
+                }
+            }
+        }
+
+    }
+
+    nonisolated private func writeAppSyncEvent(_ event: AppSyncRealTimeRequest) async throws {
+        guard await self.webSocketClient.isConnected else {
+            log.debug("[AppSyncRealTimeClient] Attempting to write to a webSocket haven't been connected.")
+            return
+        }
+
+        let interceptedEvent = await self.requestInterceptor.interceptRequest(event: event, url: self.endpoint)
+        let eventString = try String(data: Self.jsonEncoder.encode(interceptedEvent), encoding: .utf8)!
+        log.debug("[AppSyncRealTimeClient] Writing AppSyncEvent \(eventString)")
+        try await webSocketClient.write(message: eventString)
+    }
+
+    private func initAppSyncConnect() {
+        log.debug("[AppSyncRealTimeClient] Sending connectionInit message")
+        Task {
+            do {
+                try await writeAppSyncEvent(.connectionInit)
+            } catch {
+                log.debug("[AppSyncRealTimeClient] Failed to send connectInit message, error: \(error)")
+            }
+        }
+    }
+
+    private func filterAppSyncSubscriptionEvent(
+        with id: String
+    ) -> AnyPublisher<AppSyncSubscriptionEvent, Never> {
+        subject.filter { $0.id == id || $0.type == .connectionError }
```

Comment:
some comments for this func would be nice to understand why we filter on `connectionError` types as well. Is it because the event has no `id` so the filter should be on the relevant`id`  OR any `connectionError` events?

Reply:
yes, it's because of `connection_error` doesn't have an `id` field and I want to pass the connection_error to downstreams.